Additional error checking for `torch.cuda.nccl` APIs.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43247 Additional error checking for `torch.cuda.nccl` APIs.**

`torch.cuda.nccl` APIs didn't throw appropriate errors when called
with inputs/outputs that were of the wrong type and it resulted in some cryptic
errors instead.

Adding some error checks with explicit error messages for these APIs.

Differential Revision: [D23206069](https://our.internmc.facebook.com/intern/diff/D23206069/)

pritamdamania committed:
Additional error checking for `torch.cuda.nccl` APIs.

`torch.cuda.nccl` APIs didn't throw appropriate errors when called
with inputs/outputs that were of the wrong type and it resulted in some cryptic
errors instead.

Adding some error checks with explicit error messages for these APIs.

Differential Revision: [D23206069](https://our.internmc.facebook.com/intern/diff/D23206069/)

[ghstack-poisoned]

mrshenli commented on torch/cuda/nccl.py:
```diff
@@ -42,15 +42,24 @@ def init_rank(num_ranks, uid, rank):
+def _check_sequence_type(inputs):
+    if not isinstance(inputs, list) and not isinstance(inputs, tuple):
+        raise TypeError("inputs should be a list/tuple")
+
```
use 2 line spacing here?

rohan-varma commented on torch/cuda/nccl.py:
```diff
@@ -42,15 +42,24 @@ def init_rank(num_ranks, uid, rank):
 
+def _check_sequence_type(inputs):
+    if not isinstance(inputs, list) and not isinstance(inputs, tuple):
+        raise TypeError("inputs should be a list/tuple")
```
Can it also be any python iterable/list-like object? I just tried the NCCL test with `inputs = iter(inputs)` (which makes it a `list_iterable` type) and `inputs = set(inputs)`, and the current `test_all_gather` seems to pass. With this PR, would we stop supporting those types?

pritamdamania committed:
Update on "Additional error checking for `torch.cuda.nccl` APIs."


`torch.cuda.nccl` APIs didn't throw appropriate errors when called
with inputs/outputs that were of the wrong type and it resulted in some cryptic
errors instead.

Adding some error checks with explicit error messages for these APIs.

Differential Revision: [D23206069](https://our.internmc.facebook.com/intern/diff/D23206069/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM

pritamdamania committed:
Update on "Additional error checking for `torch.cuda.nccl` APIs."


`torch.cuda.nccl` APIs didn't throw appropriate errors when called
with inputs/outputs that were of the wrong type and it resulted in some cryptic
errors instead.

Adding some error checks with explicit error messages for these APIs.

Differential Revision: [D23206069](https://our.internmc.facebook.com/intern/diff/D23206069/)

[ghstack-poisoned]
